src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java

```diff
@@ -0,0 +1,190 @@
+import ru.yandex.practicum.filmorate.model.Film;
+import ru.yandex.practicum.filmorate.model.RatingsEnum;
+import ru.yandex.practicum.filmorate.storages.FilmStorage;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.time.LocalDate;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@RequiredArgsConstructor
+@Component
+@Primary
+public class FilmDbStorage implements FilmStorage {
+
+    private final JdbcTemplate jdbcTemplate;
+    private final GenresDAOImpl genresDAO;
+    private final MpaDAOImpl mpaDAO;
+
+    @Override
+    public Film addFilm(Film film) {
+        if (film.getReleaseDate().isBefore(LocalDate.of(1895, 12, 28))) {
+            throw new ValidationException("Данные в запросе на добавление нового фильма не соответствуют критериям.");
+        }
+        String sql = "INSERT INTO Film (Name, " +
+                "        Description, " +
+                "        Realise_date, " +
+                "        Duration, " +
+                "        Rating_id) values " +
+                "(?, ?, ?, ?, ?);";
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbcTemplate.update(new PreparedStatementCreator() {
+            public PreparedStatement createPreparedStatement(Connection connection) throws SQLException {
+                PreparedStatement ps = connection.prepareStatement(sql, new String[]{"Film_id"});
+                ps.setString(1, film.getName());
+                ps.setString(2, film.getDescription());
+                ps.setDate(3, java.sql.Date.valueOf(film.getReleaseDate()));
+                ps.setInt(4, film.getDuration());
+                ps.setString(5, film.getMpa().toString());
+                return ps;
+            }
+        }, keyHolder);
+        if (film.getGenres() != null) {
+            String sqlGenreAdd = "INSERT INTO Genres_line (Film_id, genre_id) " +
+                    "values " +
+                    "(?, ?)";
+            film.getGenres().forEach(g -> jdbcTemplate.update(sqlGenreAdd, keyHolder.getKey().intValue(), g.toString()));
+        }
+        return getFilmById(keyHolder.getKey().intValue());
+    }
+
+    @Override
+    public Film updateFilm(Film film) {
+        try {
+            String sql = "UPDATE FILM  SET Name = ?, Description = ?, Realise_date = ?, Duration = ?, Rating_id = ? WHERE Film_id = ?; ";
+            jdbcTemplate.update(sql,
+                    film.getName(),
+                    film.getDescription(),
+                    java.sql.Date.valueOf(film.getReleaseDate()),
+                    film.getDuration(),
+                    film.getMpa().toString(),
+                    film.getId());
+            if (film.getGenres() == null) {
+                return getFilmById(film.getId());
+            }
+            if (film.getGenres().size() == 0) {
```

Comment:
Лучше использовать метод [isEmpty()](https://www.geeksforgeeks.org/java-string-isempty-method-example/)
```suggestion
            if (film.getGenres().isEmpty()) {
```
он более читаемый